src/udev/udev-node.c

```diff
@@ -212,26 +216,50 @@ static int link_update(sd_device *dev, const char *slink, bool add) {
         if (!add && unlink(filename) == 0)
                 (void) rmdir(dirname);
 
-        r = link_find_prioritized(dev, add, dirname, &target);
-        if (r < 0) {
-                log_device_debug(dev, "No reference left, removing '%s'", slink);
-                if (unlink(slink) == 0)
-                        (void) rmdir_parents(slink, "/");
-        } else
-                (void) node_symlink(dev, target, slink);
-
         if (add)
                 do {
                         _cleanup_close_ int fd = -1;
 
-                        r = mkdir_parents(filename, 0755);
+                        r = mkdir_parents_label(filename, 0755);
                         if (!IN_SET(r, 0, -ENOENT))
                                 break;
                         fd = open(filename, O_WRONLY|O_CREAT|O_CLOEXEC|O_TRUNC|O_NOFOLLOW, 0444);
                         if (fd < 0)
                                 r = -errno;
                 } while (r == -ENOENT);
 
+        for (int i = 0; i < LINK_UPDATE_MAX_RETRIES; i++) {
+                struct stat st1 = {}, st2 = {};
+
+                if (add && stat(dirname, &st1) < 0)
+                        return -errno;
+
+                r = link_find_prioritized(dev, add, dirname, &target);
+
+                if (add) {
+                        if (stat(dirname, &st2) < 0)
+                                return -errno;
+
+                        if (!stat_inode_unmodified(&st1, &st2))
+                                continue;
+                }
+
+                if (r < 0) {
+                        log_device_debug(dev, "No reference left, removing '%s'", slink);
+                        if (unlink(slink) == 0)
+                                (void) rmdir_parents(slink, "/");
+
+                        break;
+                } else {
+                        r = node_symlink(dev, target, slink);
+                        if (r < 0) {
+                                (void) unlink(filename);
+                                break;
+                        } else if (r == 1)
+                                continue;
```

Comment:
why this check? if you remove it, then the code will be exactly the same, as this is at the end of the loop iteration anyway

Reply:
if you move second stat() check to the end of the loop iteration then the primary exit condition for the loop should be that mtime still matches. that's the key really.

Reply:
Good point, there used to be some code bellow and I didn't fix the else branch